commit 0cdbe3a025a19e1f9242a25fc80d9091787345c1
Author: Ambroz Bizjak <ambrop7@gmail.com>
Date:   Thu Feb 4 00:45:53 2016 +0100

    emac: Remove return value from emac_dev_write_end.

diff --git a/sam/drivers/emac/emac.c b/sam/drivers/emac/emac.c
index 8b5cf38..37ee4e2 100644
--- a/sam/drivers/emac/emac.c
+++ b/sam/drivers/emac/emac.c
@@ -493,7 +493,7 @@ uint32_t emac_dev_write_start(emac_device_t* p_emac_dev, uint32_t ul_size, void
 	return EMAC_OK;
 }
 
-uint32_t emac_dev_write_end(emac_device_t* p_emac_dev, uint32_t ul_size)
+void emac_dev_write_end(emac_device_t* p_emac_dev, uint32_t ul_size)
 {
 	volatile emac_tx_descriptor_t *p_tx_td;
 	Emac *p_hw = p_emac_dev->p_hw;
@@ -518,8 +518,6 @@ uint32_t emac_dev_write_end(emac_device_t* p_emac_dev, uint32_t ul_size)
 
 	/* Now start to transmit if it is still not done */
 	emac_start_transmission(p_hw);
-	
-	return EMAC_OK;
 }
 
 /**
diff --git a/sam/drivers/emac/emac.h b/sam/drivers/emac/emac.h
index 34036c3..4f8a3a3 100644
--- a/sam/drivers/emac/emac.h
+++ b/sam/drivers/emac/emac.h
@@ -1010,7 +1010,7 @@ uint32_t emac_dev_read_start(emac_device_t* p_emac_dev, emac_dev_read_state_t *s
 		uint32_t* p_size1, uint32_t *p_size2);
 void emac_dev_read_end(emac_device_t* p_emac_dev, emac_dev_read_state_t *state);
 uint32_t emac_dev_write_start(emac_device_t* p_emac_dev, uint32_t ul_size, void **out_buffer);
-uint32_t emac_dev_write_end(emac_device_t* p_emac_dev, uint32_t ul_size);
+void emac_dev_write_end(emac_device_t* p_emac_dev, uint32_t ul_size);
 uint32_t emac_dev_get_tx_load(emac_device_t* p_emac_dev);
 void emac_dev_set_rx_callback(emac_device_t* p_emac_dev,
 		emac_dev_tx_cb_t func_rx_cb);
