commit 478c95c031ede11e1d0c7f38eb3d246e33bb048c
Author: Ambroz Bizjak <ambrop7@gmail.com>
Date:   Sun Jul 31 13:36:11 2016 +0200

    Changes for APrinter.

diff --git a/Drivers/CMSIS/Include/cmsis_gcc.h b/Drivers/CMSIS/Include/cmsis_gcc.h
index d868f2e..f312ee1 100644
--- a/Drivers/CMSIS/Include/cmsis_gcc.h
+++ b/Drivers/CMSIS/Include/cmsis_gcc.h
@@ -337,7 +337,11 @@ __attribute__( ( always_inline ) ) __STATIC_INLINE void __set_FPSCR(uint32_t fps
 #if (__FPU_PRESENT == 1U) && (__FPU_USED == 1U)
   /* Empty asm statement works as a scheduling barrier */
   __ASM volatile ("");
+#ifdef __clang__
+  __builtin_arm_set_fpscr(fpscr);
+#else
   __ASM volatile ("VMSR fpscr, %0" : : "r" (fpscr) : "vfpcc");
+#endif
   __ASM volatile ("");
 #endif
 }
diff --git a/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c b/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c
index b57c1cd..c2cfbae 100644
--- a/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c
+++ b/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c
@@ -57,6 +57,7 @@
 
 /* Includes ------------------------------------------------------------------*/
 #include "stm32f4xx_hal.h"
+#include <string.h>
 
 /** @addtogroup STM32F4xx_LL_USB_DRIVER
   * @{
@@ -783,7 +784,9 @@ HAL_StatusTypeDef USB_WritePacket(USB_OTG_GlobalTypeDef *USBx, uint8_t *src, uin
     count32b =  (len + 3U) / 4U;
     for (i = 0U; i < count32b; i++, src += 4U)
     {
-      USBx_DFIFO(ch_ep_num) = *((__packed uint32_t *)src);
+      uint32_t val;
+      memcpy(&val, src, sizeof(val));
+      USBx_DFIFO(ch_ep_num) = val;
     }
   }
   return HAL_OK;
@@ -809,8 +812,8 @@ void *USB_ReadPacket(USB_OTG_GlobalTypeDef *USBx, uint8_t *dest, uint16_t len)
   
   for ( i = 0U; i < count32b; i++, dest += 4U )
   {
-    *(__packed uint32_t *)dest = USBx_DFIFO(0U);
-    
+    uint32_t val = USBx_DFIFO(0U);
+    memcpy(dest, &val, sizeof(val));
   }
   return ((void *)dest);
 }
diff --git a/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Inc/usbd_cdc.h b/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Inc/usbd_cdc.h
index 31397d2..206445f 100644
--- a/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Inc/usbd_cdc.h
+++ b/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Inc/usbd_cdc.h
@@ -104,7 +104,7 @@ typedef struct _USBD_CDC_Itf
   int8_t (* DeInit)        (void);
   int8_t (* Control)       (uint8_t, uint8_t * , uint16_t);   
   int8_t (* Receive)       (uint8_t *, uint32_t *);  
-
+  void   (* TxCompleted)   (void);
 }USBD_CDC_ItfTypeDef;
 
 
@@ -159,6 +159,8 @@ uint8_t  USBD_CDC_SetRxBuffer        (USBD_HandleTypeDef   *pdev,
 uint8_t  USBD_CDC_ReceivePacket      (USBD_HandleTypeDef *pdev);
 
 uint8_t  USBD_CDC_TransmitPacket     (USBD_HandleTypeDef *pdev);
+
+uint8_t  USBD_CDC_CheckTxBusy        (USBD_HandleTypeDef *pdev);
 /**
   * @}
   */ 
diff --git a/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Src/usbd_cdc.c b/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Src/usbd_cdc.c
index 06b3fc8..8b49a7d 100644
--- a/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Src/usbd_cdc.c
+++ b/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Src/usbd_cdc.c
@@ -669,6 +669,8 @@ static uint8_t  USBD_CDC_DataIn (USBD_HandleTypeDef *pdev, uint8_t epnum)
   {
     
     hcdc->TxState = 0;
+    
+    ((USBD_CDC_ItfTypeDef *)pdev->pUserData)->TxCompleted();
 
     return USBD_OK;
   }
@@ -874,6 +876,21 @@ uint8_t  USBD_CDC_TransmitPacket(USBD_HandleTypeDef *pdev)
 }
 
 
+uint8_t  USBD_CDC_CheckTxBusy(USBD_HandleTypeDef *pdev)
+{      
+  USBD_CDC_HandleTypeDef   *hcdc = (USBD_CDC_HandleTypeDef*) pdev->pClassData;
+  
+  if(pdev->pClassData != NULL)
+  {
+    return hcdc->TxState == 0 ? USBD_OK : USBD_BUSY;
+  }
+  else
+  {
+    return USBD_FAIL;
+  }
+}
+
+
 /**
   * @brief  USBD_CDC_ReceivePacket
   *         prepare OUT Endpoint for reception
