diff --git a/teensy3/mk20dx128.c b/teensy3/mk20dx128.c
index 578e703..2d82c69 100644
--- a/teensy3/mk20dx128.c
+++ b/teensy3/mk20dx128.c
@@ -350,9 +350,10 @@ extern void rtc_set(unsigned long t);
 
 
 
-static void startup_unused_hook(void) {}
-void startup_early_hook(void)		__attribute__ ((weak, alias("startup_unused_hook")));
-void startup_late_hook(void)		__attribute__ ((weak, alias("startup_unused_hook")));
+static void startup_default_early_hook(void) { WDOG_STCTRLH = WDOG_STCTRLH_ALLOWUPDATE; }
+static void startup_default_late_hook(void) {}
+void startup_early_hook(void)		__attribute__ ((weak, alias("startup_default_early_hook")));
+void startup_late_hook(void)		__attribute__ ((weak, alias("startup_default_late_hook")));
 
 
 __attribute__ ((section(".startup")))
@@ -364,7 +365,6 @@ void ResetHandler(void)
 
 	WDOG_UNLOCK = WDOG_UNLOCK_SEQ1;
 	WDOG_UNLOCK = WDOG_UNLOCK_SEQ2;
-	WDOG_STCTRLH = WDOG_STCTRLH_ALLOWUPDATE;
 	startup_early_hook();
 
 	// enable clocks to always-used peripherals
